activesupport/lib/active_support/cache.rb

```diff
@@ -790,12 +815,75 @@ def save_block_result_to_cache(name, options)
 
     module NullCoder # :nodoc:
       class << self
+        def dump(entry)
+          entry
+        end
+
         def load(payload)
           payload
         end
+      end
+    end
 
+    MARSHAL_PREFIX = "\x04\b".b.freeze # :nodoc:
+
+    module LegacyCoder # :nodoc:
+      class << self
         def dump(entry)
-          entry
+          Marshal.dump(entry)
+        end
+
+        def dump_compressed(entry, threshold)
+          Marshal.dump(entry.compressed(threshold))
+        end
+
+        def load(payload)
+          if payload.start_with?(MARSHAL_PREFIX)
+            Marshal.load(payload)
+          else
+            EntryCoder.load(payload)
+          end
+        end
+      end
+    end
+
+    module EntryCoder # :nodoc:
+      UNCOMPRESSED = 0
```

Comment:
If you read these constants, you think: Oh, these are integers. You only realize reading the code that they are supposed to be bytes. However, that's kind of a mismatch to me, because we only use them as strings of length 1, and are calling `ord` and `chr` all the time.

What do you think about making them strings?

```ruby
# Must one ASCII char.
UNCOMPRESSED = "0"
COMPRESSED = "1"
```

Then, the reader finds natural that the values are used as they are:

```ruby
UNCOMPRESSED + Marshal.dump(entry.pack)
COMPRESSED + compressed_payload
prefix = payload.byteslice(0)
case prefix
when UNCOMPRESSED
...
end
```

and the code gets simpler. I believe string allocations are on par.

Reply:
> What do you think about making them strings?

I'm fine with making them strings. However I'd rather keep them as "binary" rather than ASCII, as it's easier to ensure we won't reuse `4.chr` which is `Marshal` prefix.

Reply:
Alternatively I can also express the constants as `0x00` and `0x01`, which makes it more obvious we're dealing with bytes.

Reply:
What about

```ruby
# Must be a string of one byte.
UNCOMPRESSED = "\x00"
```

?

Reply:
Sure. I'll update the PR tomorrow.

Reply:
Done in 35d5ddd